core/types/receipt.go

```diff
@@ -87,29 +100,35 @@ func (r *Receipt) DecodeRLP(s *rlp.Stream) error {
 	if err != nil {
 		return err
 	}
-	list, _, err := rlp.SplitList(raw)
+	content, _, err := rlp.SplitList(raw)
 	if err != nil {
 		return err
 	}
-	items, err := rlp.CountValues(list)
+	kind, cnt, _, err := rlp.Split(content)
 	if err != nil {
 		return err
 	}
-	// Deserialize based on the number of content items
-	switch items {
-	case 3:
-		// Metropolis receipts have 3 components
+	// Deserialize based on the first component type.
+	switch {
+	case kind == rlp.Byte || kind == rlp.String && len(cnt) == 0:
```

Comment:
I think this part is correct, since 0x01 is decoded as a byte, but 0x80 as an empty string (0 number). It's a bit weird, but that's the way RLP guesses the type unless it's given explicitly.

One thing that did occur to me however, is that we don't enforce the status byte to be 0 or 1. That's a possible consensus problem. I'll add a fix for this.